### ISSUE TITLE
Make the AxiosResponse and AxiosPromise generic so that the type of the data can very

### ISSUE THREAD

**Issue by ofirgeller** [axios/axios#1112]
#### Summary

The user code usually expects a specific type for the data returned from a request. currently however that metadata is lost and when using AxiosResponse and AxiosPromise we have to do casting. 

Here is how you could allow setting the type while not breaking backward compatibility:

redefine the types:

```
export interface AxiosResponse<T = any> {
  data: T;
  status: number;
  statusText: string;
  headers: any;
  config: AxiosRequestConfig;
}

export interface AxiosPromise<TData = any> extends Promise<AxiosResponse<TData>> {
}


export interface AxiosInstance {
/// ...redacted code
  request<T = any>(config: AxiosRequestConfig): AxiosPromise<T>;
  get<T = any>(url: string, config?: AxiosRequestConfig): AxiosPromise<T>;
  delete<T = any>(url: string, config?: AxiosRequestConfig): AxiosPromise<T>;
  head<T = any>(url: string, config?: AxiosRequestConfig): AxiosPromise<T>;
  post<T = any>(url: string, data?: any, config?: AxiosRequestConfig): AxiosPromise<T>;
  put<T = any>(url: string, data?: any, config?: AxiosRequestConfig): AxiosPromise<T>;
  patch<T = any>(url: string, data?: any, config?: AxiosRequestConfig): AxiosPromise<T>;
}
```

Now you can do this and get an error

```
let promise = axios.request<{email:string}>(config);
promise.then(res=>{
    console.log('The user email is ' + res.data.eail); /// ERROR
});
```

I can look at the actual code (not the d.ts files) and make a pull request if you want.

#### Context

- axios version: 1.6
- typescript >= 2.3 (for the default generic type)

**Comment by d-fischer** [comment]
Duplicate of #718 - also already fixed in #1061 but nobody merges it

**Comment by nickuraltsev** [comment]
Fixed via #1061
